contracts/STYLE_GUIDE.md

```diff
@@ -0,0 +1,347 @@
+
+### Sending tokens
+
+- Prefer [ERC20.safeTransfer](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#SafeERC20) over ERC20.transfer
+
+### Gas golfing
+
+- Golf your code. Make it cheap, within reason.
+  - Focus on the hot path
+  - Check the [Solidity guild page](https://www.notion.so/Solidity-Guild-642cfdcfdb0a403c9a8fdb5015ff8244?pvs=21) for recordings on gas golfing
+- Most of the cost of executing Solidity is related to reading/writing storage
+- Calling other contracts will also be costly
+- Common types to safely use are
+  - uint40 for timestamps (or uint32 if you really need the space)
+  - uint96 for link, as there are only 1b link tokens
+- prefer `++i` over `i++`
+- If you’re unsure about golfing, ask in the #tech-solidity channel
+
+# Testing
+
+- Test using Foundry, do not use Hardhat
+- Aim for 90%+ *useful* coverage as a baseline, but (near) 100% is achievable in Solidity. Always 100% test the critical path.
+  - Make sure to test for each event emitted
+  - Test each reverting path
+- Consider fuzzing, start with stateless (very easy in Foundry) and if that works, try stateful fuzzing.
+- Consider fork testing if applicable
+
+### Foundry
+
+- Create a Foundry profile for each project folder in `foundry.toml`
+- Foundry tests live in the project folder in `src`, not in the `contracts/test/` folder
+- Set the block number and timestamp. It is preferred to set these values to some reasonable value close to reality.
+- There should be no code between `vm.expectEmit`/`vm.expectRevert` and the function call
+
+# Picking a Pragma
+
+- If a contract or library is expected to be imported by outside parties then the pragma should be kept as loose as possible without sacrificing safety. We publish versions for every minor semver version of Solidity, and maintain a corresponding set of tests for each published version.
+  - Examples: libraries, interfaces, abstract contracts, and contracts expected to be inherited from
+- Otherwise, Solidity contracts should have a pragma which is locked to a specific version.
+  - Example: Most concrete contracts.
+- Avoid changing pragmas after audit. Unless there is a bug that has affects your contract, then you should try to stick to a known good pragma. In practice this means we typically only support one (occasionally two) pragma for any “major”(minor by semver naming) Solidity version.
+- The current advised pragma is `0.8.19` or higher, lower versions should be avoided when starting a new project. Newer versions can be considered.
```

Comment:
When starting a new project, is picking the latest version advisable? Is there merit in picking an older, more battle tested version?

Reply:
The latest that is already used in CL products would be default imo. The absolute latest (given it has been out for 3+ months) can be considered. There should be no reason so start any project with <0.8.19 as multiple products already use it. If it wasn't safe, we would not be using it